src/test/java/org/jenkinsci/plugins/envinject/EnvInjectPluginActionTest.java

```diff
@@ -0,0 +1,153 @@
+        
+        assertEquals("Injected variables should be hidden", EnvInjectVarList.HIDDEN, target);
+        assertFalse("Users should have no permission to see injected vars", 
+                canViewInjectedVars(User.getUnknown(), build));
+    }
+    
+    @Test
+    public void hideActionViaPermissions() throws Exception {
+        // Enable permissions
+        final EnvInjectPlugin plugin = EnvInjectPlugin.getInstance();
+        EnvInjectPluginConfiguration.configure(false, true);
+        
+        // Create a test user
+        User user = User.get("testUser", true, null);
+             
+        final GlobalMatrixAuthorizationStrategy strategy = new GlobalMatrixAuthorizationStrategy();
+        strategy.add(Jenkins.READ, user.getId());
+        strategy.add(Item.READ, user.getId());
+        strategy.add(Item.DISCOVER, user.getId());
+        j.jenkins.setAuthorizationStrategy(strategy);
+        
+        // Run build and retrieve results
+        FreeStyleBuild build = j.buildAndAssertSuccess(p);
+        assertFalse("User should have no View Injected Vars permission", 
+                hasPermission(user, build, EnvInjectPlugin.VIEW_INJECTED_VARS));
+        assertFalse("User should have no permission to see injected vars", 
+                canViewInjectedVars(user, build));
+            
+        // Grant permissions and check the results 
+        strategy.add(EnvInjectPlugin.VIEW_INJECTED_VARS, user.getId());
+        assertTrue("User should have the View Injected Vars permission", 
+                hasPermission(user, build, EnvInjectPlugin.VIEW_INJECTED_VARS));
+        assertTrue("User should have a permission to see injected vars", 
+                canViewInjectedVars(user, build));
+    }
+    
+    
+    private boolean hasPermission(User user, AccessControlled item, Permission permission ) 
+            throws AssertionError {
+        SecurityContext initialContext = null;
+        try {
+            Authentication auth = user.impersonate();
+            initialContext = hudson.security.ACL.impersonate(auth);
+            return item.hasPermission(permission);
+        } catch (UsernameNotFoundException ex) {
+            throw new AssertionError("The specified user is not found", ex);
+        } finally {
+            if (initialContext != null) {
+                SecurityContextHolder.setContext(initialContext);
+            }
+        }
+    }
+    
+    private boolean canViewInjectedVars(@Nonnull User user, @Nonnull Run<?,?> run) 
+            throws UsernameNotFoundException {
+        SecurityContext initialContext = null;
+        try {
+            Authentication auth = user.impersonate();
+            initialContext = hudson.security.ACL.impersonate(auth);
+            return EnvInjectPlugin.canViewInjectedVars(run);
+        } finally {
+            if (initialContext != null) {
+                SecurityContextHolder.setContext(initialContext);
+            }
+        }
```

Comment:
BTW can be written more concisely with an effectively final variable:

```java
Authentication auth = user.impersonate();
SecurityContext initialContext = ACL.impersonate(auth);
try {
    return EnvInjectPlugin.canViewInjectedVars(run);
} finally {
    SecurityContextHolder.setContext(initialContext);
}
```

(There are also two newer overloads with closures, but neither are suitable for Java 8 lambdas.)